pkg/gardenlet/controller/shoot/shoot_control_delete.go

```diff
@@ -492,10 +492,15 @@ func (r *shootReconciler) runDeleteShootFlow(ctx context.Context, o *operation.O
 			Dependencies: flow.NewTaskIDs(destroyControlPlane),
 		})
 
+		waitUntilShootManagedResourcesDeleted = g.Add(flow.Task{
+			Name:         "Waiting until shoot managed resources have been deleted",
+			Fn:           flow.TaskFn(botanist.WaitUntilShootManagedResourcesDeleted).DoIf(cleanupShootResources).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(waitUntilControlPlaneDeleted),
+		})
 		deleteKubeAPIServer = g.Add(flow.Task{
 			Name:         "Deleting Kubernetes API server",
 			Fn:           flow.TaskFn(botanist.DeleteKubeAPIServer).RetryUntilTimeout(defaultInterval, defaultTimeout),
-			Dependencies: flow.NewTaskIDs(syncPointCleaned, waitUntilControlPlaneDeleted),
+			Dependencies: flow.NewTaskIDs(syncPointCleaned, waitUntilControlPlaneDeleted).InsertIf(cleanupShootResources, waitUntilShootManagedResourcesDeleted),
```

Comment:
I think you don't need the `InsertIf` here. You can directly add the dependency. If the `DoIf` call for waitUntilShootManagedResourcesDeleted evaluates to false, the flow graph will automatically consider that `waitUntilShootManagedResourcesDeleted` has been completed (there's an [`EmptyTaskFn`](https://github.com/gardener/gardener/blob/a36dbb178c844a28904af4570eaaa485bcc4299e/pkg/utils/flow/taskfn.go#L39) that gets executed in this case)

```suggestion
			Dependencies: flow.NewTaskIDs(syncPointCleaned, waitUntilControlPlaneDeleted, waitUntilControlPlaneDeleted),
```